keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,382 @@
+
+#### OpenStack
+
+The OpenStack CPI supports TCP, UDP, HTTP(S) in Service definitions and can configure the Octavia listeners with the protocols defined in the Service.
+OpenStack Octavia supports TCP, UDP and HTTP(S) on listeners, an own listener must be configured for each protocol, and different listeners can be used on the same LB instance.
+
+There was a bug in Octavia versions <5.0.0: it was not possible to use the same port number (5e.g. 53) with different protocols (e.g. TCP and UDP) on the same LB instance. It has been fixed in 5.0.0, which is available since the "T" release of OpenStack.
+
+Summary: the OpenStack based clouds that use Octavia v2 as their LBaaS seems to support this feature once implemented. It is true that the "T" release is the newest one, so upgrade may take a while. On the other hand OpenStack documentation mentions, that a newer Octavia version can be used with previous releases of other OpenStack projects, i.e. it can be the case that the upgrade effort is on Octavia side in an OpenStack cloud. 
+Pricing is up to the pricing model of the OpenStack providers.
+
+#### Oracle Cloud
+
+The Oracle CPI does not support UDP in the K8s Service definition. It supports only TCP and HTTP. It supports mixed TCP and HTTP ports in the same Service definition.
+
+The Oracle Cloud Load Balancer supports TCP, HTTP(S) protocols. 
+
+The pricing is based on time and capacity: https://www.oracle.com/cloud/networking/load-balancing-pricing.html I.e. the amount of protocols on a sinlge LB instance does not affect the pricing.
+
+Summary: Oracle CPI and LB seem to support mixed protocol Services, and the pricing is not based on the number of protocols per Service.
+
+#### Tencent Cloud
+
+The Tencent Cloud CPI supports TCP, UDP and HTTP in Service definitions. It maps "HTTP" protocol value in a Service definition to "TCP" before creating a listener on the CLB.
+The Tencent Cloud CPI can manage both of their LB solutions: "Classic CLB" and "Cloud Load Balancer" (previously known as "Application Load Balancer"). 
+The Tencent Cloud CLB supports TCP, UDP and HTTP(S) listeners, an own listener must be configured for each protocol. 
+The number of listeners does not affect CLB pricing. CLB pricing is time (day) based and not tied to the number of listeners.
+https://intl.cloud.tencent.com/document/product/214/8848
+
+A user can ask for an internal Load Balancer via a K8s Service definition that  has the annotation `service.kubernetes.io/qcloud-loadbalancer-internal-subnetid: subnet-xxxxxxxx`. Internal CLBs are free.
+
+Summary: Tencent Cloud CPI and LBs seem to support mixed protocol Services, and the pricing is not based on the number of protocols per Service.
+
+
+### Risks and Mitigations
+
+#### Billing perspective
+
+The goal of the current restriction on the K8s API was to prevent an unexpected extra charging for Load Balancers that were created based on Services with mixed protocol definitions.
```

Comment:
Maybe mention that earlier; I was kind of confused when you suddenly starting talking about billing. I had assumed that the problem was solely going to be that some clouds supported multiprotocol and others didn't (eg like AWS)

Reply:
Yes. Because as I learnt, the main reason to disable this feature explicitly was to prevent surprising/shocking doubled LB bills in clouds where the billing happens per protocol. That is why the main direction was to show how such LBs would be charged at the different cloud providers if this feature were re-enabled.

Reply:
Fixed.